### ISSUE TITLE
TiledMenu Suggestions

### ISSUE THREAD

**Issue by BigDaddyonTech** [Zren/plasma-applets#45]
I just want to say that I love the TiledMenu and appreciate all the hard work that goes into it. I made a video about the menu that you can watch here
https://www.youtube.com/watch?v=JfreWHt_bDY&lc=z12gfhsj4umuedkth22gy3p5znabtnvfa

I was wondering if in future versions it is possible to add the following options.
1. Make the search findings default to applications and have the other category follow after. For example, when I type in Firefox in the search box, the application is the top result all the way up until I enter the "x" in firefox and then the search findings switches to a firefox folder location as the top finding. In the video I explain it better.
2. When adding new shortcuts to the tile menu. Allow for inserting the item which would push all the rest of the items to the right or down to the next location rather than flipping exact locations, which then all icons need to be relocated if you already had multiple favorites set.
3. Add a setting and the ability to uncheck the "recent applications" section at the top of the menu. In the application dashboard this is a setting.

Again, I thank you for your hard work and hopefully what I have suggested does not come across that I don't appreciate the menu. They are meant as suggestions to help make a great menu even better.
Thanks for your time.

**Comment by Zren** [comment]
Awesome! Thanks for the video review.

So I actually started working on some of these (and [replying on the kde store](https://i.imgur.com/5cXb1ID.png)), but Black Friday kinda distracted me heh. I'll reply here since it actually supports text formatting...

1. The search preferences is a bit of a hack. Plasma doesn't expose the "category" (app/file/run command) when I order [the results "merged"](https://i.imgur.com/icMMobb.png) (so no category separators). It also doesn't provide the "relevance" property that KRunner uses to sort when the results aren't merged (but it then tells you what category it's in). No matter if it was merged or not, I couldn't get the exact same list as KRunner had, so I just decided to move "full text matches" to the top (which is why it prefers the folder instead of the app. It obviously needs work.
    * I'll try to create a toggle for showing the results categorized (where I try to manually sort the given results), or showing a [simple "results" list](https://i.imgur.com/icMMobb.png) that is the same as "Application Launcher".

2. Yeah, I need to insert when adding new items, might do that tonight since the drop code is still fresh in my memory.

3. Mhmm. Probably should support Recent Contacts and Files too but meh. "Recent Apps" was a compromise since Plasma doesn't have an api for the "most used" like Win10, though tbh I prefer it most recent now since "most used" is usually favourited.

From the Video

* v6 has App Descriptions "After" the name by default.
* Thanks for pointing out the white text on white bg search box bug. Fixed it.
* Just fixed the bit about being unable to drop to the "empty tile area".

Again, thanks for doing a review of my widget!

**Comment by BigDaddyonTech** [comment]
Hey bro,
Thanks for taking the time to make and update the menu. I for one really appreciate what goes into it. Thank you as well for looking at my suggestions. I understand there are limitations and not everything can be changed or needs to be changed.
One suggestion :) I forgot to mention was on the favorite tiles, It would be nice to be able to add .desktop shortcuts along with folder locations. For example, I have Linux document that I use often and I have a .desktop shortcut to it that I made with a custom icon that I put in my plank dock. It would be nice to have that in the favorite tiles.
I might have to do a follow-up video on the changes you made already

**Comment by BigDaddyonTech** [comment]
the white font fix and the dropping of the icons onto the blank part work perfectly now. Thank you

**Comment by cinnamoner** [comment]
@Zren You're fantastic! Massive Congratulations!

**Comment by Zren** [comment]
> I forgot to mention was on the favorite tiles, It would be nice to be able to add .desktop shortcuts along with folder locations.

You can drag the file the desktop file is linking to from dolphin to link to it. Or pin it's from the search results (eventually you'll be able to drag from the results / app list since that feels more natural).

You won't be able to drag a `.desktop` file though, and this is a minor problem with how my code works atm. Right now when you drop the "mime" data for looks like `file:///home/chris/test.desktop`, but if you drop an app shortcut from the menu (which currently disabled because of a bug), the url get's converted from `dolphin.desktop` to `file:///home/chris/.local/share/plasma/plasmoids/com.github.zren.tiledmenu/contents/ui/dolphin.desktop`. Note how it's prefixed with the location of the where the widget is installed? In order for the app button to work when dropped, I needed to convert the desktop url back to the first. My solution was to [look for `.desktop` at the end of the url](https://github.com/Zren/plasma-applets/blob/master/tiledmenu/package/contents/ui/FavouritesView.qml#L265). I hadn't thought people would drag custom desktop files to the menu, so I'd better fix that to only strip out the current location (somehow).

**Comment by BigDaddyonTech** [comment]
Thank you very much
I did a search for it and pinned it to the tiles. It does not have my icon on it but at least I can get to it.
Thanks again

**Comment by michaldybczak** [comment]
Thanks to BigDaddy, I also checked the menu and it's great but has some small shortcomings that can be improved in the feature:

- no localisation (other menus are localized dependent on system language, this one stays in English), I can help with Polish language
- when having a panel on the left, choosing tiled menu makes the kickstarter icon smaller, much smaller then the rest of icons which looks weird
- tiles seem to be always greyed out on a desktop theme option, luckily we can manually probe the color and change it so that's not a big issue

Ideas: changing size of tiles, changing size of a menu window by dragging it from the corner (just like on the windows 10)
Thanks!

**Comment by Zren** [comment]
> tiles seem to be always greyed out on a desktop theme option

I make config options I've yet to finish disabled rather than hidden. I sometimes plan future configuration like that when mocking up the config window, then write the feature later.

> localisation

I've avoided doing this on all my widgets so far because I don't know how to do it properly. As far as I can tell, I need to have the user run a script after installing to compile the translations... Eg: [plasma-applet-weather-widget](https://github.com/kotelnik/plasma-applet-weather-widget/tree/master/translations)

> small icon

I was mainly trying to make it follow the [same size/padding windows](https://www.youtube.com/watch?v=XrC7I1_jRA4#t=3m32) had when I wrote that bit. I was hoping to solve the problem where it scales too big when the panel is really huge.

![](https://i.imgur.com/gk3YwZF.png) ![](https://i.imgur.com/JTF5tyw.png)

It seems using the same size/padding as Windows probably not be a very good default since we're not contrained to 40px/60px thick panels. So I added a toggle which'll be off by default.

![](https://i.imgur.com/wDWbXz1.png)

I was trying to fix the issue where the icon gets really big when the panel is very thick.

-----

* Made it insert to the hovered spot when adding a new thingy
* Added option to choose merged search results (same as app launcher) or separate into categories (app menu/dash). The old custom category sorting is disabled for the near future as it's a bit wonky.

**Comment by Zren** [comment]
> changing size of a menu window by dragging it from the corner (just like on the windows 10)

You can Alt+Right click and drag to resize the menu. It sorta snaps to the correct size, but it does bug out if you keep holding without moving.

https://streamable.com/p4dy

The minor glitch is because it's a hack rather than telling the "menu window" that it's resizable. Some day I'll copy the Popup Window code so that I can set the window flags.

**Comment by BigDaddyonTech** [comment]
I gotta tell ya, You are awesome. 
This latest version you added primo features to it. "show recent applications" toggle. :)
and many others. Thank you for the continued work on it.

**Comment by michaldybczak** [comment]
> I was trying to fix the issue where the icon gets really big when the panel is very thick.

Ah, I get it. You are right. When I first configured side panel, kickstarter icon was so huge, it was annoying. However I got used to it and because it's in the same size as the rest of program icons, it's not that bad. When Using your menu, it became visibly smaller then everything which started to look weird and not proportional. Maybe I would be happy with it when I first set the panel that way? ;)

> I've avoided doing localisation on all my widgets so far because I don't know how to do it properly. As far as I can tell, I need to have the user run a script after installing to compile the translations

Then I guess it's a topic for future improvements. Check how other localised menus did that. Implementing localisation would really make it look more polished and professional. Now when I hit alternatives, your menu stands out with its English name. It's not a problem at all, at least not for me, but at some point localisation will have to be taken into account if one wants to go more pro with own packages. Luckily there is not much to translate here but still, it's a work to do. Anyway, don't rush into quick solutions, take your time with it. It's better to get it right.
